src/ioctls/device.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fs::File;
+use std::io;
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+
+use kvm_bindings::kvm_device_attr;
+
+use ioctls::Result;
+use kvm_ioctls::KVM_SET_DEVICE_ATTR;
+use sys_ioctl::ioctl_with_ref;
+
+/// Wrapper over the file descriptor obtained when creating an emulated device in the kernel.
+pub struct DeviceFd {
+    fd: File,
+}
+
+impl DeviceFd {
+    /// See the documentation for `KVM_SET_DEVICE_ATTR`.
+    pub fn set_device_attr(&self, nr_irqs_attr: &kvm_device_attr) -> Result<()> {
+        let ret = unsafe { ioctl_with_ref(self, KVM_SET_DEVICE_ATTR(), nr_irqs_attr) };
+        if ret != 0 {
+            return Err(io::Error::last_os_error());
+        }
+        Ok(())
+    }
+}
+
+/// Helper function for creating a new device.
+pub fn new_device(dev_fd: File) -> DeviceFd {
+    DeviceFd { fd: dev_fd }
+}
+
+impl AsRawFd for DeviceFd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd.as_raw_fd()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ioctls::system::Kvm;
+    use std::io::{Error, ErrorKind};
+
+    #[test]
+    fn test_create_device() {
+        let kvm = Kvm::new().unwrap();
+        let vm = kvm.create_vm().unwrap();
+
+        let mut gic_device = kvm_bindings::kvm_create_device {
```

Comment:
nit: can you import the structures and constants that you need from the kvm_bindings instead of using them with kvm_bindings::KVM_CREATE_DEVICE_TEST? The code is a bit hard to read.